cw/cw305/capture.py

```diff
@@ -330,6 +330,124 @@ def capture_aes_fvsr_key(ot):
 
+def capture_aes_fvsr_key_batch(ot, ktp, capture_cfg, scope_type):
+    """A generator for capturing AES traces for fixed vs random key test in batch mode.
+    The data collection method is based on the derived test requirements (DTR) for TVLA:
+    https://www.rambus.com/wp-content/uploads/2015/08/TVLA-DTR-with-AES.pdf
+    The measurements are taken by using either fixed or randomly selected key in batches.
+    In order to simplify the analysis, the first encryption has to use fixed key.
+    To generate random keys and texts internal PRNG's are used instead of AES functions as specified
+    in the DTR.
+
+    Args:
+      ot: Initialized OpenTitan target.
+      ktp: Key and plaintext generator.
+      capture_cfg: Capture configuration.
+      scope_type: cw or waverunner as a scope for batch capture.
+    """
+    # Seed host's PRNG.
+    # TODO: Replace this with a dedicated PRNG to avoid other packages breaking our code.
+    random.seed(capture_cfg["batch_prng_seed"])
+    # Seed the target's PRNG
+    ot.target.simpleserial_write("s", capture_cfg["batch_prng_seed"].to_bytes(4, "little"))
+
+    key_fixed = bytearray([0x81, 0x1E, 0x37, 0x31, 0xB0, 0x12, 0x0A, 0x78,
+                           0x42, 0x78, 0x1E, 0x22, 0xB2, 0x5C, 0xDD, 0xF9])
+    tqdm.write(f'Fixed key: {binascii.b2a_hex(bytes(key_fixed))}')
+    ot.target.simpleserial_write("t", key_fixed)
+
+    sample_fixed = 1
+
+    # Create the ChipWhisperer project.
+    project = cw.create_project(capture_cfg["project_name"], overwrite=True)
+    # Capture traces.
+    rem_num_traces = capture_cfg["num_traces"]
+    num_segments_storage = 1
+    # cw and waverunner scopes are supported for batch capture.
+    scope = SCOPE_FACTORY[scope_type](ot, capture_cfg)
+
+    with tqdm(total=rem_num_traces, desc="Capturing", ncols=80, unit=" traces") as pbar:
+        while rem_num_traces > 0:
+            # Determine the number of traces for this batch and arm the oscilloscope.
+            scope.num_segments = min(rem_num_traces, scope.num_segments_max)
+            scope.arm()
+
+            # Start batch encryption.
+            ot.target.simpleserial_write("f", scope.num_segments_actual.to_bytes(4, "little"))
+
+            # Transfer traces.
+            waves = scope.capture_and_transfer_waves()
+            assert waves.shape[0] == scope.num_segments
+            # Check that the ADC didn't saturate when recording this batch.
+            check_range(waves, ot.scope.adc.bits_per_sample)
+
+            # Compute expected outputs for checking at the end
+            plaintexts = []
+            ciphertexts = []
+            keys = []
+            for ii in range(scope.num_segments_actual):
+                if sample_fixed:
+                    key = np.asarray(key_fixed)
+                else:
+                    key = np.asarray(ktp.next()[1])
+                plaintext = np.asarray(ktp.next()[1])
+                ciphertext = np.asarray(scared.aes.base.encrypt(plaintext, key))
```

Comment:
I think you found out in the meantime that computing all ciphertexts here has a very high impact on the capture rate. On the target we're no longer concatenating the encryptions so we should probably also remove it here?

Reply:
Actually I am not sure if the high impact on the capture rate is justified. Because we have also been computing all ciphertexts in `capture_aes_random_batch()` and there we have a high capture rate as well. Is the problem maybe in how this piece of code here is written? Maybe too many copies?

Reply:
I was thinking that I can fix this in another PR with some other fixes. In `capture_aes_random_batch()` we were using the same key for all encyrptions. But in `capture_aes_fvsr_key_batch()` we are setting new key for almost every encryption. This may be the reason for that.

Reply:
Maybe these are two very stupid questions/comments but:
1)  why do you assign `np.asarray(ktp.next()[1])` to both key and plaintext?  Is there a typo for the index?
2) are you aware that  `ktp.next` might produce the same key all the time? https://forum.newae.com/t/scope-function-for-cwnano-correlation-power-analysis/2806/8 defined in `software/chipwhisperer/capture/acq_patterns/basic.py`

Reply:
When I call `ktp.next()[1]` over and over again, it generates random number each time. I am basically calling it to generate random numbers.

Reply:
yeah. I just realized that, but you were faster than my edit ;)  Thanks!

Reply:
Investigating this later on sounds good to me.